File: LeetCode_Java/hash/SubarraySumEqualsK.java
package hash;

import java.util.HashMap;
import java.util.Map;

public class SubarraySumEqualsK {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);

        int sum = 0;
        int result = 0;

        for (int num : nums) {
            sum += num;

            if (map.containsKey(sum - k)) {
                result += map.get(sum - k);
            }

            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }

        return result;
    }

    public static void main(String[] args) {
        SubarraySumEqualsK solution = new SubarraySumEqualsK();

        int[] nums1 = {1, 1, 1};
        int k1 = 2;
        System.out.println(solution.subarraySum(nums1, k1)); // Output: 2

        int[] nums2 = {1, 2, 3};
        int k2 = 3;
        System.out.println(solution.subarraySum(nums2, k2)); // Output: 2

        int[] nums3 = {-1, -1, 1};
        int k3 = 0;
        System.out.println(solution.subarraySum(nums3, k3)); // Output: 1
    }
}

File: LeetCode_Python/prefixsum/SubarraySumEqualsK.py
class SubarraySumEqualsK:
    def subarraySum(self, nums: list[int], k: int) -> int:
        prefix = 0
        count = {0: 1}
        res = 0

        for num in nums:
            prefix += num
            diff = prefix - k
            if diff in count:
                res += count[diff]
            count[prefix] = count.get(prefix, 0) + 1

        return res


if __name__ == "__main__":
    solution = SubarraySumEqualsK()

    print("Excepcted Output : 2")
    print("Actual Output : ", solution.subarraySum([1, 1, 1], 2))

    print("Excepcted Output : 2")
    print("Actual Output : ", solution.subarraySum([1, 2, 3], 3))

    print("Excepcted Output : 1")
    print("Actual Output : ", solution.subarraySum([1, 2, 3], 6))
